homeassistant/components/growatt_server/coordinator.py

```diff
@@ -0,0 +1,215 @@
+            )
+        except json.decoder.JSONDecodeError:
+            _LOGGER.error("Unable to fetch data from Growatt server")
+
+        return self.data
+
+    def get_currency(self):
+        """Get the currency."""
+        return self.data.get("currency")
+
+    def get_data(self, entity_description):
+        """Get the data."""
+        variable = entity_description.api_key
+        api_value = self.data.get(variable)
+        previous_value = self.previous_values.get(variable)
+        return_value = api_value
+
+        # If we have a 'drop threshold' specified, then check it and correct if needed
+        if (
+            entity_description.previous_value_drop_threshold is not None
+            and previous_value is not None
+            and api_value is not None
+        ):
+            _LOGGER.debug(
+                (
+                    "%s - Drop threshold specified (%s), checking for drop... API"
+                    " Value: %s, Previous Value: %s"
+                ),
+                entity_description.name,
+                entity_description.previous_value_drop_threshold,
+                api_value,
+                previous_value,
+            )
+            diff = float(api_value) - float(previous_value)
+
+            # Check if the value has dropped (negative value i.e. < 0) and it has only
+            # dropped by a small amount, if so, use the previous value.
+            # Note - The energy dashboard takes care of drops within 10%
+            # of the current value, however if the value is low e.g. 0.2
+            # and drops by 0.1 it classes as a reset.
+            if -(entity_description.previous_value_drop_threshold) <= diff < 0:
+                _LOGGER.debug(
+                    (
+                        "Diff is negative, but only by a small amount therefore not a"
+                        " nightly reset, using previous value (%s) instead of api value"
+                        " (%s)"
+                    ),
+                    previous_value,
+                    api_value,
+                )
+                return_value = previous_value
+            else:
+                _LOGGER.debug(
+                    "%s - No drop detected, using API value", entity_description.name
+                )
+
+        # Lifetime total values should always be increasing, they will never reset,
+        # however the API sometimes returns 0 values when the clock turns to 00:00
+        # local time in that scenario we should just return the previous value
+        if entity_description.never_resets and api_value == 0 and previous_value:
+            _LOGGER.debug(
+                (
+                    "API value is 0, but this value should never reset, returning"
+                    " previous value (%s) instead"
+                ),
+                previous_value,
+            )
+            return_value = previous_value
+
+        self.previous_values[variable] = return_value
+
+        return return_value
```

Comment:
Okay so since this already was here before I am fine with this, however, I would recommend to keep the logic on how entities get their data in the platform and not in the coordinator, as the coordinator is for fetching data and doesn't care about how entities show it.

But since this was already there, I would recommend to look at this in a followup